server/controllers/paypal.js

```diff
@@ -16,193 +15,121 @@ const {
-      paymentMethod.save()
-        .then(paymentMethod => cb(null, paymentMethod))
-        .catch(cb);
-    }]
-
-  }, (err, results) => {
-    if (err) return next(err);
-    res.json(results.callPaypal);
-  });
-
+  const startDate = new Date();
+  const endDate = (req.query.endingDate && (new Date(req.query.endingDate))) || moment().add(1, 'years');
+  
+  const payload = {
+    currencyCode: 'USD', // TODO: figure out if there is a reliable way to specify correct currency for a HOST.
+    startingDate: startDate.toISOString(),
+    endingDate: endDate.toISOString(),
+    returnUrl,
+    cancelUrl,
+    displayMaxTotalAmount: false,
+    feesPayer: 'SENDER',
+    maxAmountPerPayment: 2500.00, // PayPal claims this can go up to $10k without needing additional permissions from them.
+    clientDetails: req.remoteUser.id
+  };
+
+  let response;
+
+  return paypalAdaptive.preapproval(payload)
+  .tap(r => response = r)
+  .then(response => PaymentMethod.create({
+    service: 'paypal',
+    UserId: req.remoteUser.id,
+    token: response.preapprovalKey,
+    startDate,
+    endDate
+  }))
+  .then(() => res.json(response))
+  .catch(next);
 };
 
 /**
  * Confirm a preapproval.
- * TODO: remove async and simplify with promises #promisify
  */
 export const confirmPreapproval = function(req, res, next) {
+  let paymentMethod;
+
+  // fetch original payment method
+  return PaymentMethod.findOne({
+    where: {
+      service: 'paypal',
+      UserId: req.remoteUser.id,
+      token: req.params.preapprovalkey
+    }
+  })
+  .tap(pm => paymentMethod = pm)
+  .then(pm => pm ? 
+    Promise.resolve() : 
+    Promise.reject(new errors.NotFound('This preapprovalKey does not exist.')))
+
+  .then(() => getPreapprovalDetailsAndUpdatePaymentMethod(req.params.preapprovalkey, req.remoteUser.id, paymentMethod))
+
+  .then(() => Activity.create({
+    type: 'user.paymentMethod.created',
+    UserId: req.remoteUser.id,
+    data: {
+      user: req.remoteUser.minimal,
+      paymentMethod
+    }
+  }))
+  
+  // clean any old payment methods
+  .then(() => PaymentMethod.findAll({
+    where: {
+      service: 'paypal',
+      UserId: req.remoteUser.id,
+      token: {$ne: req.params.preapprovalkey}
+    }
+  }))
+  // TODO: Call paypal to cancel preapproval keys before marking as deleted.
+  .then(oldPMs => oldPMs && oldPMs.map(pm => pm.destroy()))
+
+  .then(() => res.send(paymentMethod.info))
+  .catch(next)
+};
 
-  async.auto({
-
-    getPaymentMethod: [function(cb) {
-      PaymentMethod
-        .findAndCountAll({
-          where: {
-            service: 'paypal',
-            UserId: req.remoteUser.id,
-            token: req.params.preapprovalkey
-          }
-        })
-        .then(paymentMethod => cb(null, paymentMethod))
-        .catch(cb);
-    }],
+/*
+ * Takes a Payment Response
+ */
+const getPreapprovalDetailsAndUpdatePaymentMethod = function(preapprovalKey, userId, paymentMethod = null) {
```

Comment:
why not splitting this in two? `getPreaprovalDetails` and `updatePaymentMethod` ?